src/types/ArraySchema.ts

```diff
@@ -94,12 +103,31 @@ export class ArraySchema<T=any> extends Array<T> {
     }
 
     filter(callbackfn: (value: T, index: number, array: T[]) => unknown, thisArg?: any): ArraySchema<T> {
-        const filtered = super.filter(callbackfn);
+        const keepItems = Array.prototype.filter.call(this, (item, idx) => {
+          return callbackfn(item, idx, this);
+        })
+
+        const removedItems = Array.prototype.filter.call(this, (item, idx) => {
+          return !callbackfn(item, idx, this);
+        })
+
+        removedItems.forEach(updatedRemovedItem);
```

Comment:
As the `ArraySchema` mimics how regular `Array`s work on JavaScript, I think it may be confusing if the `.filter()` method applies changes to the current array instance. The `.filter()` method on JavaScript just creates a new array filtering with filtered data, without mutating the array being filtered. (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)